File: Behavioral/Visitor/visitor.cs
using System;
using System.Collections.Generic;

// Interfaz del Visitor
interface IVisitor
{
    void VisitCirculo(Circulo circulo);
    void VisitCuadrado(Cuadrado cuadrado);
}

// Interfaz para elementos que aceptan visitantes
interface IFigura
{
    void Accept(IVisitor visitor);
}

// Clases concretas
class Circulo : IShape
{
    public double Radio { get; set; } = 5;
    public void Accept(IVisitor visitor) => visitor.VisitCirculo(this);
}

class Cuadrado : IShape
{
    public double Lado { get; set; } = 3;
    public void Accept(IVisitor visitor) => visitor.VisitCuadrado(this);
}

// Visitor concreto
class CalculadoraAreas : IVisitor
{
    public void VisitCirculo(Circulo circulo)
    {
        double area = Math.PI * circulo.Radio * circulo.Radio;
        Console.WriteLine($"Área del círculo: {area}");
    }

    public void VisitCuadrado(Cuadrado cuadrado)
    {
        double area = cuadrado.Lado * cuadrado.Lado;
        Console.WriteLine($"Área del cuadrado: {area}");
    }
}

// Programa principal
class Program
{
    static void Main()
    {
        List<IShape> figuras = new List<IShape> { new Circulo(), new Cuadrado() };
        CalculadoraAreas calculadora = new CalculadoraAreas();

        foreach (var figura in figuras)
        {
            figura.Accept(calculadora);
        }
    }
}

File: Behavioral/Visitor/visitor.py
import math

# Interfaz del Visitor
class Visitor:
    def visit_circulo(self, circulo):
        pass

    def visit_cuadrado(self, cuadrado):
        pass

# Clases de elementos
class Figura:
    def accept(self, visitor: Visitor):
        pass

class Circulo(Figura):
    def __init__(self, radio=5):
        self.radio = radio

    def accept(self, visitor: Visitor):
        visitor.visit_circulo(self)

class Cuadrado(Figura):
    def __init__(self, lado=3):
        self.lado = lado

    def accept(self, visitor: Visitor):
        visitor.visit_cuadrado(self)

# Visitor concreto
class CalculadoraAreas(Visitor):
    def visit_circulo(self, circulo: Circulo):
        area = math.pi * circulo.radio ** 2
        print(f"Área del círculo: {area}")

    def visit_cuadrado(self, cuadrado: Cuadrado):
        area = cuadrado.lado ** 2
        print(f"Área del cuadrado: {area}")

# Uso
figuras = [Circulo(), Cuadrado()]
calculadora = CalculadoraAreas()

for figura in figuras:
    figura.accept(calculadora)
